functionary/prompt_template/base_template.py

```diff
@@ -11,6 +12,257 @@
+                    "func_name": "",
+                    "param_names": [],
+                }
+                gen_state["stage"] = "function"
+        elif gen_state["stage"] == "function":
+            gen_state["func_name"] = gen_state["curr_text"].rstrip()
+
+            # Check if the new state is in "pre-parameter" stage
+            if gen_state["curr_text"].endswith(
+                self.get_stop_token_for_function_parameter(stage="function")
+            ) or (
+                sum([gen_state["curr_text"] == option for option in options]) == 1
+                and sum(
+                    [option.startswith(gen_state["curr_text"]) for option in options]
+                )
+                == 1
+            ):
+                gen_state["stage"] = "pre-parameter"
+
+                # Update curr_text and curr_tokens
+                if gen_state["curr_text"].endswith(
+                    self.get_stop_token_for_function_parameter(stage="function")
+                ):
+                    gen_state["curr_text"] = self.get_stop_token_for_function_parameter(
+                        stage="function"
+                    )
+                    gen_state["curr_tokens"] = [new_token_id]
+                else:
+                    gen_state["curr_text"], gen_state["curr_tokens"] = "", []
+        elif gen_state["stage"] == "pre-parameter":
+            # Check if the new state is in "parameter" stage
+            if self.fn_param_sep_token in gen_state["curr_text"]:
+                gen_state["stage"] = "parameter-name"
+                gen_state["curr_text"], gen_state["curr_tokens"] = "", []
+        elif gen_state["stage"] == "parameter-name":
+            # Get the latest param
+            latest_param_str = gen_state["curr_text"]
+
+            # Check if the new state is in "parameter-value" stage
+            stop_token = self.get_stop_token_for_function_parameter(stage="parameter")
+            if latest_param_str.endswith(stop_token):
+                gen_state["param_names"].append(
+                    gen_state["curr_text"].removesuffix(stop_token)
+                )
+                gen_state["stage"] = "parameter-value"
+                gen_state["curr_text"] = stop_token
+                gen_state["curr_tokens"] = [new_token_id]
+        elif gen_state["stage"] == "parameter-value":
+            # Check if the new state is in "pre-function" stage
+            try:
+                _ = json.loads(
+                    '{"' + gen_state["param_names"][-1] + gen_state["curr_text"]
+                )
+                gen_state["stage"] = "pre-function"
+            except:
+                pass
+
+            # Check if the current state can be converted to json, it means the
+            # new state is back to "parameter-name" stage
+            try:
+                _ = json.loads(
+                    '{"'
+                    + gen_state["param_names"][-1]
+                    + gen_state["curr_text"].removesuffix(', "')
```

Comment:
I just wonder if:
{"x": 123456}
and 123456 is splitted into: "123", "456", then it will stop at: {"x": 123 ? and remove to new state ?

Reply:
gotcha. Will add somemore checks here.

Reply:
```python
if gen_state["curr_text"].endswith(', "'):
    """Conduct the json.loads operation"""
```